GSB Docstrings, Sample Files and Tests

cczhu opened the pull request:
Implemented all necessary 'gsb/base.py' docstrings (those that are user-facing), and revised any other docs for self-consistency.

Created GSB sample files from GMRT observations of the Crab Pulsar and of PSR J1810+1744 (both obtained by Rob), and placed them in `baseband/data`.  Rewrote `gsb/tests/test_gsb.py` to use this new data.  Annoying it's not possible to place the sample files in a subfolder of `data` for cleanliness - Astropy doesn't copy them into `tmpdir` when testing.  Is there a way around this issue?

During implementation, made two code modifications:
- `sub_int` was renamed to `mem_block` (Jayanta Roy says the term is actually a redundant modulo-8 shared memory block number used during telescope-side debugging) and its redundancy noted in the docs.
- `size` is now a property of GSB headers (and assumes 1 trailing space and 1 carriage return at the end of each header line) and called by `seek_offset`.

fixes #62

mhvk commented:
@cczhu - on subdirectories in `data` - it may be that you have to add them explicitly to the main `setup.py` (though I'm not sure any more... check astropy too)
```
package_info['package_data'][PACKAGENAME].append('data/*')
```

mhvk commented on baseband/gsb/header.py:
```diff
@@ -292,10 +322,11 @@ def seek_offset(self, n, size=None):
         size : int, optional
             The size in bytes of the present header (if not given, it will
-            be calculated assuming the termination string is a single CR).
+            be calculated assuming the termination string is a space followed
```
As above

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -206,58 +207,118 @@ def test_decoding(self):
+        # Test encoding then decoding consistency via `fromdata`.
+        payload7 = gsb.payload.GSBPayload.fromdata(payload4.data.real, bps=8)
+        assert np.all(payload7.data == payload4.data.real)
+        # Wrong BPS will flatten high payload values.
```
I don't understand how the next test can work!?

cczhu replied:
I think (it's been a while since I wrote this) that this just checks that if an incorrect bps is passed, `fromdata` uses `encode_4bit` rather than `encode_8bit` and you lose any values that can't be expressed in 4 bits.  Since it's more confusing than useful (and the encoders are tested elsewhere) I'll delete it.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -206,58 +207,118 @@ def test_decoding(self):
+            assert payload4.sample_shape == (1,)
+            assert payload4.sample_shape.nchan == 1
+            payload5 = gsb.GSBPayload.fromdata(payload4.data, bps=8)
+            assert np.all(payload5.data == payload4.data)
```
I think it would make more sense to show that `payload5.words == payload4.words` here.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -206,58 +207,118 @@ def test_decoding(self):
+            #   2 complex elements/channel * 1 bytes/element)
+            payload4 = gsb.GSBPayload.fromfile(fh, payloadsize=2**12, bps=8,
+                                               complex_data=True)
+            # Values from reading pol0, thread0 phased file as if it were raw
```
From here on one can be out of the open-file context.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -206,58 +207,118 @@ def test_decoding(self):
+            payload8 = gsb.payload.GSBPayload.fromdata(payload4.data.real,
+                                                       bps=4)
+            assert np.all(payload8 == payload4.data.real)
+        # Encode complex data.
```
How is this test different from that with `payload5`?

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -270,241 +331,286 @@ def test_rawdump_frame(self, tmpdir):
+
+        with gsb.open(SAMPLE_RAWDUMP_HEADER, mode='rs', raw=SAMPLE_RAWDUMP,
+                      samples_per_frame=2**13) as fh_r, \
+                io.open(str(tmpdir.join('test_time.timestamp')), 'w+b') as sh,\
```
no need for `io.`

cczhu replied:
In Python 2 removing `io` gives:

```
 AttributeError: 'file' object has no attribute 'readable'
```

which I thought was why the old `test_gsb.py` also only opened timestamp files with `io`?

mhvk replied:
If it is a python2 thing, can we remove the `io.` here, but add at the top
```
if six.PY2:
    from io import open
```
That way we can remove the clutter when we're python3-only (not too long from now...)

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -270,241 +331,286 @@ def test_rawdump_frame(self, tmpdir):
 
         # Test that opening that raises an exception correctly handles
         # file closing. (Note that the timestamp file always gets closed).
         with io.open(str(tmpdir.join('test.timestamp')), 'w+b') as sh, \
```
no need for `io.`

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -270,241 +331,286 @@ def test_rawdump_frame(self, tmpdir):
-            assert np.all(data[:twopol.shape[0]] == twopol)
-            assert np.all(data[twopol.shape[0]:] == twopol[::-1])
+            sh.seek(0)
+            for pol in fraw:
```
now seen this so many times that I wonder if the test class cannot use a little helper function!

cczhu commented on baseband/data/sample_gsb_phased.timestamp:
```diff
@@ -0,0 +1,10 @@
+2013 07 28 02 53 57 0.027501 2013 07 28 02 53 56 0.834058240 10001 1 
+2013 07 28 02 53 57 0.279274 2013 07 28 02 53 57 0.085716480 10002 2 
+2013 07 28 02 53 57 0.530805 2013 07 28 02 53 57 0.337374720 10003 3 
+2013 07 28 02 53 57 0.782605 2013 07 28 02 53 57 0.589032960 10004 4 
```
Timestamp files don't have newlines (though it's hard to tell with phased, since all the ones I have access to end in incomplete headers), so I'll keep this as is.

mhvk replied:
That suggests perhaps we should put in an incomplete header - that will test `stop_time` better as well.

cczhu commented:
I've now removed explicit `io` calls in `test_gsb.py`, and implemented the conditional import you suggested.

Also, following your point about checking if the last header is complete or not, I wrote a try/except block under `GSBStreamReader`'s `_last_header` routine.  The assert statement is to catch instances where a rawdump header is only missing one or two char of the fractional second, and therefore will successfully be read in by `GSBRawdumpHeader`.  The try block is clunky, though, and I would be happy to entertain alternatives.  Even if we keep `_last_header` as it once was, we can keep the tests I wrote to check if an incomplete header will return an error.

mhvk commented:
@cczhu - I wonder if we're approaching the last header the wrong way. E.g., we could also search based on how many blocks there are in the raw data file. Could you check on the data we have available whether that would point one to the last-but-one header? Certainly, we should check that on an existing file, we can do `fh.size`... (which needs the last header).

cczhu commented:
@mhvk - The raw files often have dozens of additional frames' worth of data, even in cases where the timestamp file does not end in a partial header.  I'm reticent to provide an automatic fix for files with missing data or header information, even if we threw a warning when doing so.

We do use `fh.size` a few times, but never directly check that it's equal to 10 payload sizes. I've now added that test to `test_gsb.py`.

mhvk commented:
OK, so there is no direct correspondence... I'm still wondering whether it is most reasonable to throw an exception, or to just take the last complete header (presumably just before). E.g., one could seek back 2.5 headers and do the split on `\n`. But it is also fine to leave this for another issue.

cczhu commented:
@mhvk I would prefer that whenever a file is broken the user must go make the repair themselves, to prevent Baseband from inadvertently obfuscating a data error, so prefer throwing an error over silently fixing it or throwing a warning.  Though maybe my intuition is off and we should just conservatively dump any data that could be incorrect because we have so much data already?

mhvk commented:
In general, I agree, but since you wrote that all GSB timestamp files you looked at had incomplete last lines, I think maybe we should be a bit more generous...

mhvk commented:
p.s.  Also, note that we will need to eventually write code that can deal with gaps, etc., in which case this would just constitute a very small gap at the end.

cczhu commented:
Sorry, I should have been exact in that statement.  Of the four phased observations Rob gave me access to, two have incomplete last headers, and payloads that contain least 30 additional frames beyond the last header.  One has a complete last header, but also a payload that extends 40 frames beyond the last header.  I know of one rawdump timestamp with a missing last header, but couldn't find the corresponding raw file, and the two complete rawdump/payload file sets I have access to are self-consistent.  So this timestamp file problem is very common, but not ubiquitous.

mhvk commented:
Ubiquitous enough, it seems to me, that we might as well guard against it.

cczhu commented:
I've revised `_last_header` to properly handle incomplete last lines (now taking advantage of the fact that we store the right size in the header instance).
